airflow/cli/api/client.py

```diff
@@ -0,0 +1,228 @@
+)
+from airflow.utils.types import DagRunType
+from airflow.version import version
+
+if TYPE_CHECKING:
+    # # methodtools doesn't have typestubs, so give a stub
+    def lru_cache(maxsize: int | None = 128):
+        def wrapper(f):
+            return f
+
+        return wrapper
+else:
+    from methodtools import lru_cache
+
+log = structlog.get_logger(logger_name=__name__)
+
+__all__ = [
+    "Client",
+    "Credentials",
+]
+
+
+def add_correlation_id(request: httpx.Request):
+    request.headers["correlation-id"] = str(uuid7())
+
+
+def get_json_error(response: httpx.Response):
+    """Raise a ServerResponseError if we can extract error info from the error."""
+    err = ServerResponseError.from_response(response)
+    if err:
+        log.warning("Server error ", extra=dict(err.response.json()))
+        raise err
+
+
+def raise_on_4xx_5xx(response: httpx.Response):
+    return get_json_error(response) or response.raise_for_status()
+
+
+def noop_handler(request: httpx.Request) -> httpx.Response:
+    path = request.url.path
+    log.debug("Dry-run request", method=request.method, path=path)
+    # TODO change for test
+    if path.startswith("/task-instances/") and path.endswith("/run"):
+        # Return a fake context
+        return httpx.Response(
+            200,
+            json={
+                "dag_run": {
+                    "dag_id": "test_dag",
+                    "run_id": "test_run",
+                    "logical_date": "2021-01-01T00:00:00Z",
+                    "start_date": "2021-01-01T00:00:00Z",
+                    "run_type": DagRunType.MANUAL,
+                },
+            },
+        )
+    return httpx.Response(200, json={"text": "Hello, world!"})
+
+
+# Credentials for the API
+class Credentials:
+    """Credentials for the API."""
+
+    api_url: str
+    api_token: str
+
+    def __init__(self, api_url: str, api_token: str = "NO_TOKEN"):
+        self.api_url = api_url
+        self.api_token = api_token
+
+    def save(self):
+        """Save the credentials."""
+        default_config_dir = os.path.expanduser("~/.config/airflow")
+        print(f"Saving credentials to {default_config_dir}")
+        if not os.path.exists(default_config_dir):
+            os.makedirs(default_config_dir)
+        with open(os.path.join(default_config_dir, "credentials.json"), "w") as f:
+            json.dump(
+                {"api_url": self.api_url, "api_token": base64.b64encode(self.api_token.encode()).decode()}, f
+            )
+
+    @classmethod
+    def load(cls) -> Credentials:
+        """Load the credentials."""
+        default_config_dir = os.path.expanduser("~/.config/airflow")
+        if os.path.exists(default_config_dir):
+            with open(os.path.join(default_config_dir, "credentials.json")) as f:
+                credentials = json.load(f)
+            return cls(
+                api_url=credentials["api_url"],
+                api_token=base64.b64decode(credentials["api_token"]).decode(),
```

Comment:
We should design the credentials file so that it can cope with multiple airflow installed concurrently -- for instance a stage and a prod deployment.

Reply:
Conformed, the user can also switch between environments such as stag, prod or so using environment variables. They can set many environments as they want. I will create an issue and improve how to make this switch in CLI which will come in another PR. Thanks!